wordpress-mobile/WordPress-Android
# [Dependency Analysis] Add Android Gradle Plugin

Project Thread: paaHJt-6yU-p2
Required By: [BuildkiteCI#456](https://github.com/Automattic/buildkite-ci/pull/456)

This PR adds the dependency analysis Android Gradle plugin to this project, for dependency analysis purposes. Also, a custom meta-data is added to the dependency analysis scheduled build so that it can be easily queried via Buildkite's REST API.

-----

For now, only the main `buildHealth` task is going to be utilized and produce data once every week on CI (see this [commit](https://github.com/wordpress-mobile/WordPress-Android/pull/21022/commits/65e421a9e7dcf0c9625850a01eae45a5debea80e) and [this](https://github.com/Automattic/buildkite-ci/pull/456) PR). Amongst other, this data will include the following:

- Number of modules (`projectCount`)
- Unused dependencies which should be removed (`unusedCount`)
- Transitive dependencies which should be declared directly (`undeclaredCount`)
- Existing dependencies which should be modified to be as indicated (`misDeclaredCount`)
- Dependencies which could be compile-only (`compileOnlyCount`)
- Dependencies which should be removed or changed to runtime-only (`runtimeOnlyCount`)

Afterward, this data will get collected from CI and uploaded to our Apps Metrics infrastructure, for visualization and alerting purposes.

-----

## To Test:



1. Local: Run the `./gradlew buildHealth` task and verify that under the root level `build/reports/dependency-analysis` folder you get the below 2 reports both, in JSON and text format:
    - `build-health-report.json`
    - `build-health-report.txt`
2. CI: Using the `New Build` 🟢 CI button for [WPAndroid](https://buildkite.com/automattic/wordpress-android/), test this standalone dependency analysis job (see form below). Then:
    - When CI starts, make sure that adding `/meta-data` to that CI build's URL ([example](https://buildkite.com/automattic/wordpress-android/builds/19136/meta-data)) will give you one extra meta-data, the `schedule-build` one, with a value of `dependency-analysis` (see screenshot below):
    - When CI completes, make sure that the below 2 CI artifacts are available both, in JSON and text format:
        - `build/reports/dependency-analysis/build-health-report.json`
        - `build/reports/dependency-analysis/build-health-report.txt`
3. REST: Using the Buildkite's REST API you could query for this type of builds, using the [meta-data](https://buildkite.com/docs/apis/rest-api/builds#list-all-builds) query parameter, and see that the result is the expect one. Use this query: `https://api.buildkite.com/v2/organizations/automattic/pipelines/wordpress-android/builds?meta_data[scheduled-build]=dependency-analysis`

![image](https://github.com/wordpress-mobile/WordPress-Android/assets/9729923/f3eea8ae-efe3-4b3d-a4f7-d9de3c56bdf1)

![image](https://github.com/wordpress-mobile/WordPress-Android/assets/9729923/ea02deb5-cb22-451c-be0f-73464d88d766)

-----

## Regression Notes: (`N/A`)

-----

## PR Submission Checklist:


-----

## Testing Checklist (strike-out the not-applying and unnecessary ones): `N/A`
